### ISSUE TITLE
Generic experiment, run for training job, notebook run

### ISSUE THREAD

**Issue by Jeffwan** [kubeflow/kubeflow#4955]
/kind feature

**Why you need this feature:**
Currently, only KFP has the experiment, run concept. In traditional ML system, it's pretty common to have these concepts for all workloads. We can consider katib-job, tf-job, notebook run, pipeline as available runners/workloads. 

The reason we want to have this is we need a carrier to aggregate artifacts, tracking, metrics for that run. This help a lot on Monitoring, MLops area. It also helps a lot on kubeflow/metadata, kubeflow/training-operator adoption. 

Ideal CUJ is DS can choose to run a notebook, tf/pytorch/katib job, or pipeline. metadata tracks algorithm metrics, we may consider to gather container level usage metric as well. Aggregate everything for that run somewhere. 

**Describe the solution you'd like:**
I think my ask is if it's worth to have a separate component aims to solve these kind of problems. I feel it's hard to ask user to move everything to KFP to leverage existing KFP capabilities. I know a few users just use low level operator directly. 

This is very rough discussion on current limitations, I am trying to collect more feedbacks and come up a detail proposal if we think it's on the right track. 

**Anything else you would like to add:**
/cc @jlewi @animeshsingh

**Comment by animeshsingh** [comment]
> I think my ask is if it's worth to have a separate component aims to solve these kind of problems. I feel it's hard to ask user to move everything to KFP to leverage existing KFP capabilities. I know a few users just use low level operator directly.
> 
> This is very rough discussion on current limitations, I am trying to collect more feedbacks and come up a detail proposal if we think it's on the right track.

Well we can assume that concepts of "Experiments" which is a collection of "Jobs/Runs" is generic enough - and the fact that implementation for it happens to be in Kubeflow Pipelines is orthogonal. Good thing is that UI exists/API exists, and now its up to the different components  (TF-Job/PyTorch/Katib) etc. to take advantage of it, and standardize around it (without thinking that its a "core" KFP functionality only.

**Comment by jlewi** [comment]
+100 to @Jeffwan 's idea that we shouldn't force people to use KFP just to collect metadata.

The idea that we explored in the past was having the concept of having a logger that would collect information and write to an approriate store.  The analogy here is how logging works in Kubernetes.

The idea is we should define a flexible pattern/convention by which applications can surface information that should be logged. e.g. we could define certain annotations for input/output etc...

We can then have agents which look for these patterns and log information to different backends (ML Metadata, MLFlow, Weights and Biases etc...).

@zhenghuiwang implemented an initial prototype for metadata
https://github.com/kubeflow/metadata/tree/master/watcher

I don't think a single "Experiment" resource will work for ML anymore then a single Resource works in kubernetes.

I think like Kubernetes we need to define a [resource model](https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/) which is a set of patterns that all objects are expected to follow but around which resources can be further customized.

As a starting point

* Here are the [4 types of objects](https://www.kubeflow.org/docs/components/metadata/#learn-more-about-the-metadata-sdk) currently defined according to metadata
* [MLFlow concepts](https://mlflow.org/docs/latest/tracking.html#concepts)


@aronchick has also been doing a lot of investigation into a shared list of nouns. Any links to add?

/cc @zhenghuiwang

**Comment by yosit** [comment]
Today, kfp an katib are two separated components. I was surprised to see that when katib runs a trial it is not shown in the experiments in kfp. I would argue that there is a duplication of capabilities between the two that needs to be consolidated. 
We should have a generic way to run an experiment/trial -that should be used by both Katib, notebooks and kfp. 
As a user regardless of how a trial/experiment is running I would like to have all the metrics, logs, checkpoints, parameters, and even the data to examine further if needed.

**Comment by animeshsingh** [comment]
+1. I have been advocating for it for long. That's how most of the Data and ML projects are architected top-down, and services which are similar should be made common across components, which means one experiment tracking engine, one concept of "jobs/runs", and a common metadata engine.

**Comment by jlewi** [comment]
Does 
> As a user regardless of how a trial/experiment is running I would like to have all the metrics, logs, checkpoints, parameters, and even the data to examine further if needed.

imply 
> We should have a generic way to run an experiment/trial -that should be used by both Katib, notebooks and kfp.

I agree metrics and logs should be consolidated in a central place. I'm not sure that means we should have a "generic" way to run hyperparameter tuning and pipelines.

Pipelines and Hyperparameter tuning are different enough that the corresponding resources should be different; e.g. with an HP job you need to specify a search space.

Borrowing a page from Kubernetes. I think we need an "Experiment model" similar  to the [Kubernetes resource model](https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/). So just like Kubernetes has a plethora of different types of resources we should have a plethora of Experiment types. But all experiment types should share certain characteristics/attributes in order to enable bulk operations and data collection.

/cc @aronchick

**Comment by aronchick** [comment]
+1 to @jlewi - we should have a broad structure we all agree on, but then let specific frameworks define whatever they want inside it

**Comment by yosit** [comment]
> Pipelines and Hyperparameter tuning are different enough that the corresponding resources should be different; e.g. with an HP job you need to specify a search space.
> 
> /cc @aronchick

When a hyperparameter runs an instance of on trial - it could essentially launch a pipeline with parameters. 
A pipeline could include one step to train a model but could possibly have multi-step with data prep or take into account the parameters when querying for data and then do a training session.

At the end of the day - an instance of a trial is just executing something we document (parameters, code, metrics, logs, and data).

Since a pipeline contains steps and each step requires different resources it only makes sense that any other plugin for kubeflow wouldn't need to implement the same functionality (launching a process and track it).

**Comment by animeshsingh** [comment]
we can go broad and duplicate 80% of similar concepts, or we can reconcile and reuse. Concept of experiments/runs/metadata should be common and reused wherever possible, including Katib and KFP

**Comment by jlewi** [comment]
> we can go broad and duplicate 80% of similar concepts, or we can reconcile and reuse. 

What does reconcilation and reuse actually look like?  Given the KFP, Katib, Job operators, etc... are all being developed by different individuals and often different organizations with different timelines and priorities; how do we go about reconciling?

**Comment by richardsliu** [comment]
Any interest in the community to revive this discussion? Seems like having a metadata storage is one of the top requested features: https://blog.kubeflow.org/kubeflow-continues-to-move-to-production#user-requests
